tsl/src/fdw/data_node_scan_plan.c

```diff
@@ -674,6 +688,689 @@ push_down_group_bys(PlannerInfo *root, RelOptInfo *hyper_rel, Hyperspace *hs,
+	partition_extra->inner_unique = extra->inner_unique;
+	partition_extra->sjinfo = extra->sjinfo;
+	partition_extra->semifactors = extra->semifactors;
+	partition_extra->param_source_rels = extra->param_source_rels;
+	partition_extra->mergeclause_list =
+		castNode(List, adjust_appendrel_attrs(root, (Node *) extra->mergeclause_list, 1, &appinfo));
+	partition_extra->restrictlist =
+		castNode(List, adjust_appendrel_attrs(root, (Node *) extra->restrictlist, 1, &appinfo));
+
+	return partition_extra;
+}
+
+/*
+ * Generate the paths for a pushed down join. Each data node will be considered as a partition
+ * of the join. The join can be pushed down if:
+ *
+ * (1) The setting "ts_guc_enable_per_data_node_queries" is enabled
+ * (2) The outer relation is a distributed hypertable
+ * (3) The inner relation is marked as a reference table
+ * (4) The join is a left join or an inner join
+ *
+ * The join will be performed between the multiple DataNodeRels (see function
+ * build_data_node_part_rels) and the original innerrel of the join (the reftable).
+ */
+void
+data_node_generate_pushdown_join_paths(PlannerInfo *root, RelOptInfo *joinrel, RelOptInfo *outerrel,
+									   RelOptInfo *innerrel, JoinType jointype,
+									   JoinPathExtraData *extra)
+{
+	TsFdwRelInfo *fpinfo;
+	Path *joinpath;
+	double rows = 0;
+	int width = 0;
+	Cost startup_cost = 0;
+	Cost total_cost = 0;
+	Path *epq_path = NULL;
+	RelOptInfo **hyper_table_rels;
+	RelOptInfo **join_partition_rels;
+	int nhyper_table_rels;
+	List *join_part_rels_list = NIL;
+#if PG15_GE
+	Bitmapset *data_node_live_rels = NULL;
+#endif
+
+	/*
+	 * Skip check if the join result has been considered already.
+	 */
+	if (joinrel->fdw_private)
+		return;
+
+	/*
+	 * This code does not work for joins with lateral references, since those
+	 * must have parameterized paths, which we don't generate yet.
+	 */
+	if (!bms_is_empty(joinrel->lateral_relids))
+		return;
+
+	/* The outerrel has to be distributed. This condition should be always hold
+	 * because otherwise we should not start the planning for distributed tables. However
+	 * we can check this condition here and fail more nicely by doing nothing than hitting
+	 * a failed assert. */
+	TimescaleDBPrivate *outerrel_private = outerrel->fdw_private;
+	if (outerrel_private == NULL || outerrel_private->fdw_relation_info == NULL)
+		return;
+
+	/* Ensure pushdown planning is only performed on the AN. On the DNs the outererl is not
+	 * partitioned. */
+	if (outerrel->nparts <= 0)
+		return;
+
+	/* Get the hypertable from the outer relation. */
+	RangeTblEntry *rte_outer = planner_rt_fetch(outerrel->relid, root);
+
+	/* Test that the fetched outer relation is a distributed hypertable. If not,
+	 * the pushdown is not possible. This should be already tested by the previous
+	 * checks on 'outerrel'. But fail nicely if the condition is not meet. */
+	if (rte_outer == NULL || !is_distributed_hypertable(rte_outer->relid))
```

Comment:
This should probably be an assertion. This function should never be invoked on a non-distributed hypertable.

Reply:
Changed to an assertion.

Reply:
Hmm, I don't see the assertion. I mean, the fact that his is a data node scan means it must be a distributed hypertable.

Reply:
We are called by the generic PostgreSQL `set_join_pathlist_hook` planner hook and have only checked that `fdw_private` and `fdw_relation_info` [are not null](https://github.com/timescale/timescaledb/pull/5212/files#diff-5d89399a9796f5e49176485595f03e7ce25cea608cdd8417d8d91d00c57396faR1678). However, we have not checked if the relation is a (distributed) hypertable. I changed this statement to an assert, which was triggered during my tests. Unfortunately, this only occurred once, and I was unable to reproduce this behavior again.

Therefore, I reverted this to a `return` statement; I would like to make sure that the `is_distributed_hypertable` check is also performed in production code and I would like to be safe here (e.g., we are not processing a table that has a populated `fdw_private` field from another FDW extension).